src/Keboola/StorageApi/Client.php

```diff
@@ -1305,6 +1305,168 @@ public function uploadFile($filePath, FileUploadOptions $options)
+            $logfn = function ($message) {
+                if (trim($message) != '') {
+                    $this->log($message, ['source' => 'AWS SDK PHP debug']);
+                }
+            };
+            $options['debug'] = [
+                'logfn' => function ($message) use ($logfn) {
+                    call_user_func($logfn, $message);
+                },
+                'stream_size' => 0,
+                'scrub_auth' => true,
+                'http' => true
+            ];
+        }
+
+        $s3Client = new \Aws\S3\S3Client($options);
+
+        $asyncUploadOptions = [];
+        if ($newOptions->getIsEncrypted()) {
+            $asyncUploadOptions['ServerSideEncryption'] = $uploadParams['x-amz-server-side-encryption'];
+        }
+
+        /*
+         * In case of an upload failure (\Aws\Exception\MultipartUploadException) there is no sane way of figuring out
+         * which part of which slice failed and partially restart upload for only that part.
+         * So the whole circus has to start over again.
+         */
+        do {
+            try {
+                $manifest = [
+                    'entries' => []
+                ];
+                $promises = [];
+                $fileHandles = [];
+                foreach ($slices as $filePath) {
+                    $fh = @fopen($filePath, 'r');
+                    $fileHandles[] = $fh;
+                    if ($fh === false) {
+                        throw new ClientException("Error on file upload to S3: " . $filePath, null, null, 'fileNotReadable');
+                    }
+                    $promises[] = $s3Client->uploadAsync(
+                        $uploadParams['bucket'],
+                        $uploadParams['key'] . basename($filePath),
+                        $fh,
+                        $uploadParams['acl'],
+                        ["params" => $asyncUploadOptions]
+                    );
+                    $manifest['entries'][] = [
+                        "url" => "s3://" . $uploadParams['bucket'] . "/" . $uploadParams['key'] . basename($filePath)
+                    ];
+                }
+                \GuzzleHttp\Promise\unwrap($promises);
+                $finished = true;
+                foreach ($fileHandles as $fh) {
+                    fclose($fh);
+                }
+
+                $manifestUploadOptions = [
+                    'Bucket' => $uploadParams['bucket'],
+                    'Key' => $uploadParams['key'] . 'manifest',
+                    'Body' => json_encode($manifest)
+                ];
+                if ($newOptions->getIsEncrypted()) {
+                    $manifestUploadOptions['ServerSideEncryption'] = $uploadParams['x-amz-server-side-encryption'];
+                }
+                $s3Client->putObject($manifestUploadOptions);
+            } catch (\Aws\Exception\MultipartUploadException $e) {
+                // TODO retry, log?
+            }
+        } while (!isset($finished));
```

Comment:
No, to je docela dobrá poznámka. Nevím, vycházel jsem z té jejich ukázky. Můžem to omezit na nějaký vysoký číslo, běžně se mi stávalo, že se to třeba 3x pustilo znovu.